src/serverModel.ts

```diff
@@ -1,47 +1,219 @@
+// TODO: Move to Consent Module
+export interface GDPRConsentStateDTO {
+    [key: string]: PrivacyDTO;
+}
 
-export default function ServerModel(mpInstance) {
-    var self = this;
-    function convertCustomFlags(event, dto) {
-        var valueArray = [];
-        dto.flags = {};
+// TODO: Move to Consent Module
+export interface CCPAConsentStateDTO {
+    data_sale_opt_out: PrivacyDTO;
+}
 
-        for (var prop in event.CustomFlags) {
-            valueArray = [];
+// TODO: Break this up into GDPR and CCPA interfaces
+export interface ConsentStateDTO {
+    gdpr?: GDPRConsentStateDTO;
+    ccpa?: CCPAConsentStateDTO;
+}
 
-            if (event.CustomFlags.hasOwnProperty(prop)) {
-                if (Array.isArray(event.CustomFlags[prop])) {
-                    event.CustomFlags[prop].forEach(function(
-                        customFlagProperty
+// TODO: Confirm which attributes are optional
+export interface ServerDTO {
+    id: string | number;
+    nm: string | number;
+    pr: number;
+    qt: number;
+    br: string | number;
+    va: string | number;
+    ca: string | number;
+    ps: number;
+    cc: string | number;
+    attrs: Record<string, string> | null;
+    dp_id?: string;
+    dp_v?: number;
+    n?: string;
+    et?: number;
+    ua?: Dictionary<string | string[]>;
+    ui?: SDKUserIdentity[];
+    ia?: Dictionary<Dictionary<string>>;
+    str?: IStore;
+    sdk?: string;
+    sid?: string;
+    sl?: number;
+    ssd?: number;
+    dt?: number;
+    dbg?: boolean;
+    ct?: number;
+    lc?: SDKGeoLocation;
+    o?: boolean;
+    eec?: number;
+    av?: string;
+    cgid?: string;
+    das?: string;
+    mpid?: MPID;
+    smpids?: MPID[];
+    con?: ConsentStateDTO;
+    fr?: boolean;
+    iu?: boolean;
+    at?: number;
+    lr?: string;
+    flags?: Dictionary<string[]>;
+    cu?: string;
+    sc?: {
+        pl: ProductDTO[];
+    };
+    pd?: {
+        an: number;
+        cs: number;
+        co: string;
+        pl: ProductDTO[];
+        ta: string;
+        ti: string;
+        tcc: string;
+        tr: number;
+        ts: number;
+        tt: number;
+    };
+    pm?: {
+        an: string;
+        pl: PromotionDTO[];
+    };
+    pi?: ProductImpressionDTO[];
+    pet?: number;
+}
+
+export interface ProductDTO {
+    id: string | number;
+    nm: string | number;
+    pr: number;
+    qt: number;
+    br: string | number;
+    va: string | number;
+    ca: string | number;
+    ps: number;
+    cc: string | number;
+    tpa: number;
+    attrs: Record<string, string> | null;
+}
+
+export interface PromotionDTO {
+    id: string | number;
+    nm: string | number;
+    ps: string | number;
+    cr: string;
+}
+
+export interface ProductImpressionDTO {
+    pil: string;
+    pl: ProductDTO[];
+}
+
+export interface UploadObject extends SDKEvent {
+    ClientGeneratedId: string;
+    Store: IStore;
+    ExpandedEventCount: number;
+    ProfileMessageType: number;
+}
+
+export interface IServerModel {
```

Comment:
do we have rules for using `I` as a prefix for interfaces vs not??

I think there can be readability benefits from being consistent

Reply:
I'm fine with using `I` as a prefix